C702.tex

```diff
@@ -0,0 +1,150 @@
+Columns are defined by a name, datatype, and a number of optional properties.
+These properties define the semantics of the column, such as units, UCD, and xtype, and may also specify the column's behavior in a database, such as whether it is nullable.
+Both tables and columns may define a TAP index, which is used to recommend an ordering for clients.
+
+Constraints define rules that restrict column values in a table.
+Felis supports primary key, foreign key, and unique constraints.
+Each constraint defines the columns to which it applies, as well as the target table and columns for foreign key constraints.
+
+Indexes indicate that a column or set of columns should be indexed in the database for faster query performance.
+Indexes are defined by a list of one or more columns.
+The indexes can be used as an indication to clients that they should consider using these columns in their queries for better performance.
+
+\section{Schema Validation}
+
+The YAML data defining the schema is validated when loaded, ensureing that the schema is correctly defined according to the model.
+The model comprises a set of Python classes inheriting from Pydantic's BaseModel class and corresponding to the different types of schema objects.
+The attributes of the classes are defined using Pydantic's Field class, which allows for specifying the data type, default value, and validation rules.
+This allows for strict validation of the schema, ensuring that all required fields are present and that the data types are correct.
+
+Additionally, various "business rules" are defined and enforced using Python "validator" functions with special annotations indicating whether they will validate the raw data or the generated Python objects.
+These functions run automatically as part of the validation process.
+An example business rule enforced by a validator might ensure that table names are unique within a given schema.
+Any validation errors which occur will cause the validation process to fail, preventing the schema from being loaded.
+Errors are reported to the user with helpful messages indicating what went wrong.
+
+\section{Python API}
+
+\articlefigure{C702_f2.eps}{fig:python}{Example of using the Felis Python API.}
+
+Felis provides a Python API for working with schema data.
+Fig.\,\ref{fig:python} shows a simple example of working with this API.
+The basic functionality consists of a set of classes representing the schema objects (Schema, Table, Column, etc.).
+Extension modules provides capabilities for generating DDL statements or loading the data into a TAP\_SCHEMA database.
+Future extensions are planned to support additional functionality, such as converting tabular data to other formats or performing database migrations.
+
+\section{Felis Data Types}
+
+Felis defines its own system of data types mapping to target database types.
+Columns are required to have a datatype field.
```

Comment:
```suggestion
Columns are required to have a \texttt{datatype} field.
```
maybe? ie this is the name of the field? As opposed to "data type"?

Reply:
It is the name of the field so I'll incorporate your suggestion.